File: study/ts/script/scene/view/customActivity/CustomActivityWelfareView.ts
const {ccclass, property} = cc._decorator;

import CommonStoryAvatar from '../../../ui/component/CommonStoryAvatar'

import CommonTalkNode from '../../../ui/component/CommonTalkNode'
import ViewBase from '../../ViewBase';
import { handler } from '../../../utils/handler';
import { CustomActivityUIHelper } from './CustomActivityUIHelper';
import { Colors } from '../../../init';
import { RichTextHelper } from '../../../utils/RichTextHelper';
import { RichTextExtend } from '../../../extends/RichTextExtend';

@ccclass
export default class CustomActivityWelfareView extends ViewBase {

   @property({
       type: cc.Label,
       visible: true
   })
   _textActTitle: cc.Label = null;

   @property({
       type: cc.Node,
       visible: true
   })
   _textNode: cc.Node = null;

   @property({
       type: cc.Label,
       visible: true
   })
   _textActDes: cc.Label = null;

   @property({
       type: cc.Node,
       visible: true
   })
   _countDownNode: cc.Node = null;

   @property({
       type: cc.Label,
       visible: true
   })
   _textTimeTitle: cc.Label = null;

   @property({
       type: cc.Label,
       visible: true
   })
   _textTime: cc.Label = null;

   @property({
       type: CommonTalkNode,
       visible: true
   })
   _commonTalk: CommonTalkNode = null;

   @property({
       type: CommonStoryAvatar,
       visible: true
   })
   _commonStoryAvator: CommonStoryAvatar = null;
   
   private _customActUnitData:any;
   private _refreshHandler:any;

onCreate() {
    this._commonStoryAvator.updateUI(204);
    this._commonStoryAvator.node.setScale(0.6);
}
onEnter() {
    this._startRefreshHandler();
}
onExit() {
    this._endRefreshHandler();
}
_startRefreshHandler() {
    if (this._refreshHandler != null) {
        return;
    }
    this._refreshHandler = handler(this, this._onRefreshTick);
    this.schedule(this._refreshHandler, 1);
}
_endRefreshHandler() {
    if (this._refreshHandler != null) {
        this.unschedule(this._refreshHandler);
        this._refreshHandler = null;
    }
}
_onRefreshTick(dt) {
    var actUnitdata = this._customActUnitData;
    if (actUnitdata) {
        this._refreshActTime(actUnitdata);
    }
}
_refreshActTime(actUnitData) {
    var timeStr = '';
    if (actUnitData.isActInRunTime()) {
        timeStr = CustomActivityUIHelper.getLeftDHMSFormat(actUnitData.getEnd_time());
    } else {
        timeStr = CustomActivityUIHelper.getLeftDHMSFormat(actUnitData.getAward_time());
    }
    this._textTime.string = (timeStr);
}
_refreshDes() {
    if (!this._customActUnitData) {
        return;
    }
    this._textActTitle.string = (this._customActUnitData.getSub_title());
    this._createProgressRichText(this._customActUnitData.getDesc());
}
refreshView(customActUnitData) {
    this._customActUnitData = customActUnitData;
    this._refreshDes();
    if (this._customActUnitData) {
        this._refreshActTime(this._customActUnitData);
    }
    this._commonTalk.setString(this._customActUnitData.getDetail(), 325, true, 325, 76, null, null, true);
}
_createProgressRichText(msg) {
    var richMsg = RichTextHelper.getRichMsgListForHashText(msg, Colors.BRIGHT_BG_RED, Colors.DARK_BG_TWO, 20);
    this._textNode.removeAllChildren();
    var widget = RichTextExtend.createWithContent(richMsg);
    widget.node.setAnchorPoint(new cc.Vec2(0, 1));
    // widget.ignoreContentAdaptWithSize(false);
    widget.maxWidth = 360;
    this._textNode.addChild(widget.node);
}
enterModule() {
    this._commonTalk.doAnim();
}


}

File: study/lua/src/app/scene/view/customactivity/CustomActivityWelfareView.lua
local ViewBase = require("app.ui.ViewBase")
local CustomActivityWelfareView = class("CustomActivityWelfareView", ViewBase)

function CustomActivityWelfareView:ctor()
	self._customActUnitData = nil
	self._textActTitle = nil
	self._textActDes = nil
	self._commonTalk = nil
	self._textNode = nil
	local resource = {
		file = Path.getCSB("CustomActivityWelfareView", "customactivity"),
		size = G_ResolutionManager:getDesignSize(),
		binding = {
		},
	}
	CustomActivityWelfareView.super.ctor(self, resource)
end

function CustomActivityWelfareView:onCreate()
	self._commonStoryAvator:updateUI(204)
	self._commonStoryAvator:setScale(0.6)
end

function CustomActivityWelfareView:onEnter()
	self:_startRefreshHandler()
end

function CustomActivityWelfareView:onExit()
	self:_endRefreshHandler()
end
function CustomActivityWelfareView:_startRefreshHandler()
	local SchedulerHelper = require("app.utils.SchedulerHelper")
	if self._refreshHandler ~= nil then
        return
	end
	self._refreshHandler = SchedulerHelper.newSchedule(handler(self,self._onRefreshTick),1)
end

function CustomActivityWelfareView:_endRefreshHandler()
	local SchedulerHelper = require("app.utils.SchedulerHelper")
	if self._refreshHandler ~= nil then
		SchedulerHelper.cancelSchedule(self._refreshHandler)
		self._refreshHandler = nil
	end
end


function CustomActivityWelfareView:_onRefreshTick( dt )
	local actUnitdata = self._customActUnitData
	if actUnitdata then
		self:_refreshActTime(actUnitdata)
	end


	--对比ActList决定是否刷新,这样太费性能了
	--G_UserData:getCustomActivity():getShowActUnitDataArr()
	--现在点击按钮触发判断
end

function CustomActivityWelfareView:_refreshActTime(actUnitData)
	local CustomActivityUIHelper = require("app.scene.view.customactivity.CustomActivityUIHelper")
	local timeStr = ""
	if actUnitData:isActInRunTime() then
		timeStr = CustomActivityUIHelper.getLeftDHMSFormat(actUnitData:getEnd_time())
		-- text = Lang.get("days7activity_act_end_time",{time = timeStr})
	else
		timeStr = CustomActivityUIHelper.getLeftDHMSFormat(actUnitData:getAward_time())
		-- text = Lang.get("days7activity_act_reward_time",{time = timeStr})
	end
	self._textTime:setString(timeStr)
end

function CustomActivityWelfareView:_refreshDes()
	if not self._customActUnitData then
		return
	end
	self._textActTitle:setString(self._customActUnitData:getSub_title())
	--self._textActDes:setString(self._customActUnitData:getDesc())

	self:_createProgressRichText(self._customActUnitData:getDesc())
end

function CustomActivityWelfareView:refreshView(customActUnitData)
	self._customActUnitData = customActUnitData
	self:_refreshDes()
	if self._customActUnitData then
		self:_refreshActTime(self._customActUnitData)
	end

	self._commonTalk:setString(self._customActUnitData:getDetail(),325,true,325,76,nil,nil,true)

	
end

function CustomActivityWelfareView:_createProgressRichText(msg)
	local RichTextHelper = require("app.utils.RichTextHelper")
	local richMsg =  json.encode(RichTextHelper.getRichMsgListForHashText(
				msg,Colors.BRIGHT_BG_RED,Colors.DARK_BG_TWO,20))
	self._textNode:removeAllChildren()
    local widget = ccui.RichText:createWithContent(richMsg)
    widget:setAnchorPoint(cc.p(0,1))
	widget:ignoreContentAdaptWithSize(false)
	widget:setContentSize(cc.size(360,0))--高度0则高度自适应
    self._textNode:addChild(widget)
end



function CustomActivityWelfareView:enterModule()
	
	self._commonTalk:doAnim()
end


return CustomActivityWelfareView
